### PR TITLE
RCS 2.0 - include remote indices privileges in `GetUserPrivilege`

### PR DESCRIPTION
**Pull request by n1v0lg**
This PR updates the `GetUserPrivileges` API to include `remote_indices`
privileges assigned to a given user's role.

### CODE REVIEW

**Comment by n1v0lg** [comment]
@elasticsearchmachine run elasticsearch-ci/release-tests

**Comment by n1v0lg** [comment]
@elasticsearchmachine run elasticsearch-ci/release-tests

**Review by AthenaEryma — COMMENTED** [review]
Left a comment I'd like to at least discuss before merging this, but otherwise this LGTM - good work!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponse.java
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponse.java:111]: Many responses of security APIs, including this one, do not travel across wire. So in practice how this method behaves has no impact. 

It should not be a problem for a mixed cluster because:
1. We do _not_ allow roles with remote indices to be created in a mixed cluster that has node older than 8.6
2. Reserved roles are node local. If a reserved role is updated with remote indices in v8.6. A old node, e.g. 8.5, will keep returning its old definition for this role. While a 8.6 node will return the new definition.

For point 2, it does mean a user with the said reserved role will get different privileges on nodes of different versions. But this is not new. Any changes to reserved roles today can have this effect.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/Role.java
Comment (ywangd) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/Role.java:233]: This method seems to be test only. Do we really need it?

**Comment by ywangd** [comment]
I read through the main code changes and skimmed through the tests. They look good. I don't have anything major to comment. I'll leave it to @gwbrown for approval since he is the designated reviewer. Thanks!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponse.java
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponse.java:111]: Good q! As Yang points out, this won't go on the wire since the action will be executed directly on the coordinating node (I'll send you a Slack thread for more context around local-only security actions) so the reserved roles will be node-local.

> We do not allow roles with remote indices to be created in a mixed cluster that has node older than 8.6

We enforce this [here](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativeRolesStore.java#L224-L233).

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/Role.java
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/Role.java:233]: Yup, will nuke.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponse.java
Comment (justincr-elastic) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponse.java:53]: Nit: Collections.emptySet()

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponse.java
Comment (justincr-elastic) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponse.java:111]: Thank you for the link. I was wondering how we can implement that a min version. Maybe it should be in a helper method, but not in scope for this PR.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponse.java
Comment (justincr-elastic) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponse.java:143]: Nit: Rename to `remoteIndices`? Can we avoid using same `indices` variable name in both RemoteIndices and Indices?

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java
Comment (justincr-elastic) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java:195]: Nit: Rename to randomResponseRcs2? I found overloading the method to be confusing while reading the code. It would also be unclear when writing new code using auto-complete, but appending the feature name to the method could make it clearer in PR reviews and future code.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java
Comment (justincr-elastic) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java:217]: Randomize choosing an empty set here? Or in callers to randomResponse() versus randomResponse(true)?

I am wondering if it makes sense to name the methods randomResponse() and randomResponseRcs2(), and add a helper randomResponseRcs1OrRcs2() when callers want to randomize between RCS 1.0 vs 2.0.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java
Comment (justincr-elastic) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java:47]: Will roles be allowed to mix "indices" and "remote_indices"? If yes or no, it might be useful to add a test for that scenario to ensure whatever we chose is working as designed. If someone comes along later and toggles it, a unit test would catch that issue.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponse.java
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponse.java:143]: I think `indices` makes sense here because it's an actual `Indices` instance, which `RemoteIndices` uses as a field. Renaming to `remoteIndices` would be kind of quirky because we'd have code like 

```
RemoteIndices remoteIndices = new RemoteIndices(); 
remoteIndices.remoteIndices();
```

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java
Comment (n1v0lg) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java:195]: Happy to add a `randomResponseWithRemoteIndices()` but would prefer not to introduce Rcs2 in the code. It's more of a project name than a code-level concept.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java
Comment (n1v0lg) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java:47]: Hmm I'm not positive about what you mean by mix `indices` and `remote_indices`, could you clarify?

The `Role` class has two separate fields, `indices` and `remoteIndices` so they are separate here and cannot be mixed.

Within the test, when we generate a response, the response is randomized and can have 0 or more indices and 0 or more remote indices, so we are covering cases where the response contains both indices and remote indices.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponse.java
Comment (justincr-elastic) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponse.java:143]: I am OK with that quirk. For example, it makes find easier when using non-IDEs like GitHub PRs.

If we implemented from scratch, perhaps we would have differentiated the two as LocalIndices versus RemoteIndices. The former has an implicit clusterAlias, the latter has explicit clusterAlias(es). The varable names might reflect that difference as LocalIndices.localIndices versus RemoteIndices.remoteIndices.

This is a nit. I will leave the decision to you. If you don't want to change it, that is fine. Just wanted you to consider if there is a way to make find easier in non-IDEs by possibly using unique variable names.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java
Comment (n1v0lg) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java:217]: > Randomize choosing an empty set here?

This is covered by the fact that we use `randomArray` which may be empty.

> I am wondering if it makes sense to name the methods randomResponse() and randomResponseRcs2(), and add a helper randomResponseRcs1OrRcs2() when callers want to randomize between RCS 1.0 vs 2.0.

Thanks for raising. I just realized that `randomResponse()` _should_ actually include remote indices makes the default behavior less confusing.  Now the suite uses `randomResponse()` everywhere, except one method where it's used as:

```
final Version version = VersionUtils.randomCompatibleVersion(random(), Version.CURRENT);
final boolean canIncludeRemoteIndices = version.onOrAfter(RoleDescriptor.VERSION_REMOTE_INDICES);
final GetUserPrivilegesResponse original = randomResponse(canIncludeRemoteIndices);
```  

Which is hopefully made less ambiguous by the var name. Let me know if you'd still prefer to have an explicit `randomResponseWithoutRemoteIndices` for the above test.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java
Comment (n1v0lg) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java:195]: Replying on this [below](https://github.com/elastic/elasticsearch/pull/90826#discussion_r1014227787)

**Comment by n1v0lg** [comment]
Failure is unrelated.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java
Comment (justincr-elastic) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java:47]: I think I am conflating `QC user role descriptor(s)` vs `FC user role descriptor`. I think it is safe for QC roles to include indices and remote_indices. I am going to open a new issue for `FC user role descriptor`.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponse.java
Comment (gwbrown) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponse.java:111]: Thanks! What I was missing in my initial comment was that reserved roles are per-node - I was up to date on Point 1, but that's enforce in the Put Role action, which internal processes may not go through. I'm glad you're one step ahead of me on this; I'm just trying to be extra careful because streaming problems are one of the relatively few ways to dig a hole that's really hard to get out of if/when they come up as SDHs.

**Review by AthenaEryma — APPROVED** [review]
LGTM, thanks! I see there's still a few ongoing conversations with Justin but I'm good with this.

**Comment by gwbrown** [comment]
> Failure is unrelated.

Can you check to see if there's an issue open for the test failure and link it, or create one if there isn't?

You can also ask CI to re-run the failing suites by commenting `@elasticsearchmachine run elasticsearch-ci/part-2-fips` (not in backticks, at the start of a new line IIRC).

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java
Comment (justincr-elastic) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java:47]: Resolving this thread. Having indices and remote_indices in same role is OK.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java
Comment (justincr-elastic) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java:217]: I would prefer to have different names. I will leave it up to you to decide on the names.

I would caution putting "RemoteIndices" in the name, though. There are 5 QC privileges (indices, run_as, cluster, global, applications). So far we added remote_indices, but is there a possibility we will get asked to add analogous settings for the other 4 privileges. For example, remote_run_as, remote_cluster, remote_global, or remote_applications.

Comment (justincr-elastic) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java:217]: Unrelated observation:
- The design copied `indices` to `remote_indices`, and added a list called `remote_indices.remote_clusters`.
- If we follow the same pattern for `cluster`, it ends up with a list called `remote_cluster.remote_clusters`.

Comment (n1v0lg) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponseTests.java:217]: For now, I'll stick with the current method names. `randomResponse()` is self-explanatory. To avoid redundant code, we want a private method with a signature along the lines of `randomResponse(boolean allowRemoteIndices)`. Since this is a private helper method inside a single test suite, it's going to be straightforward to rename and tweak should we ever add other types of remote privileges (for which there currently is no medium term requirement). Should this requirement come up, happy to revisit this!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponse.java
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponse.java:143]: Sounds good! Think it's one of those things where it comes down more to personal preference than anything. I'll run with `indices` to keep things moving. Since this is just a field name, as opposed to part of the REST API, we can rename if naming inspiration strikes in the future.

**Pull request merged**
